File: ruby/linked_list.rb
class Node
  attr_accessor :data, :next

  def initialize(data)
    @data = data
    @next = nil
  end
end

class LinkedList
  attr_reader :head

  def initialize(data)
    @head = Node.new(data)
  end

  def insert(data, current=@head)
    if current.next
      insert(data, current.next)
    else
      current.next = Node.new(data)
    end
  end

  def find(data, current=@head)
    if current.data == data
      current
    else
      find(data, current.next) if current.next
    end
  end

  def update(find_data, update_data, current=@head)
    node = find(find_data)
    if node
      node.data = update_data
      true
    else
      false
    end
  end

  def find_parent(data, current=@head)
    if current.next && current.next.data == data
      current
    else
      find_parent(data, current.next) if current.next
    end
  end

  def destroy(data)
    node = find(data)
    if node
      parent = find_parent(data)
      if parent
        parent.next = node.next
      else
        @head = node.next
      end
      true
    else
      false
    end
  end

  def print(current=@head)
    puts current.data
    print(current.next) if current.next
  end
end

File: python/linked_list.py

# python linked list - basic CRUD

# Create X
# Read
# Update
# Destroy

class Node:
    def __init__(self, data):
        self.data = data
        self.next = None


class LinkedList:
    def __init__(self, data):
        self.head = Node(data)

    def print_list(self, current=None):
        if current == None:
            current = self.head

        print current.data
        if current.next:
            self.print_list(current.next)

    def insert(self, data, current=None):
        if current == None:
            current = self.head
            
        if current.next == None:
            current.next = Node(data)
        else:
            self.insert(data, current.next)

    def find(self, data, current=None):
        if current == None:
            current = self.head

        if current.data == data:
            return current
        elif current.next != None:
            return self.find(data, current.next)

ll = LinkedList("first node")
ll.insert("second node")
ll.insert("third node")
# ll.print_list()
node = ll.find("third node")
print node.data
